jaas/modules/src/main/java/org/apache/karaf/jaas/modules/ldap/LDAPBackingEngine.java

```diff
@@ -0,0 +1,210 @@
+            filter = filter.replaceAll(Pattern.quote("%u"), "*");
+            filter = filter.replace("\\", "\\\\");
+
+            LOGGER.debug("Looking for the users in LDAP with ");
+            LOGGER.debug("  base DN: " + options.getUserBaseDn());
+            LOGGER.debug("  filter: " + filter);
+
+            NamingEnumeration namingEnumeration = context.search(options.getUserBaseDn(), filter, controls);
+            try {
+                while (namingEnumeration.hasMore()) {
+                    SearchResult result = (SearchResult) namingEnumeration.next();
+
+                    // We need to do the following because slashes are handled badly. For example, when searching
+                    // for a user with lots of special characters like cn=admin,=+<>#;\
+                    // SearchResult contains 2 different results:
+                    //
+                    // SearchResult.getName = cn=admin\,\=\+\<\>\#\;\\\\
+                    // SearchResult.getNameInNamespace = cn=admin\,\=\+\<\>#\;\\,ou=people,dc=example,dc=com
+                    //
+                    // the second escapes the slashes correctly.
+                    String userDNNamespace = result.getNameInNamespace();
+                    // handle case where cn, ou, dc case doesn't match
+                    int indexOfUserBaseDN = userDNNamespace.toLowerCase().indexOf("," + options.getUserBaseDn().toLowerCase());
+                    String userDN = (indexOfUserBaseDN > 0) ?
+                            userDNNamespace.substring(0, indexOfUserBaseDN) :
+                            result.getName();
+
+                    // we need to pull out the cn=, uid=, ect.. from the user name to get the actual user name
+                    String userName = userDN;
+                    if (userDN.contains("=")) userName = userDN.split("=")[1];
+                  
+                    users.add(new UserPrincipal(userName));
+
+
+                }
+            } finally {
+                if (namingEnumeration != null) {
+                    try {
+                        namingEnumeration.close();
+                    } catch (NamingException e) {
+                        // Ignore
+                    }
+                }
+            }
+
+            return users;
+
+        } catch (NamingException e) {
+            throw new RuntimeException(e);
+        }
+
+
+    }
+
+    @Override
+    public List<GroupPrincipal> listGroups(UserPrincipal user) {
+        // for now return empty list, group implementation is not supported
+        return Collections.emptyList();
+    }
+
+    @Override
+    public Map<GroupPrincipal, String> listGroups() {
+        // for now return empty list, group implementation is not supported
+        return Collections.emptyMap();
+    }
+
+    @Override
+    public void addGroup(String username, String group) {
+        throw new RuntimeException("Adding a group is not supporting in LDAP");
```

Comment:
What about `UnsupportedOperationException` instead of a generic `RuntimeException` ?